x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java

```diff
@@ -0,0 +1,148 @@
+        Environment environment,
+        ActionFilters actionFilters) {
+        super(KibanaEnrollmentAction.NAME, transportService, actionFilters, KibanaEnrollmentRequest::new);
+        this.environment = environment;
+        // Should we use a specific origin for this ? Are we satisfied with the auditability of the change password request as-is ?
+        this.client = new OriginSettingClient(client, SECURITY_ORIGIN);
+        this.settings = settings;
+        this.sslService = sslService;
+    }
+
+    @Override protected void doExecute(Task task, KibanaEnrollmentRequest request, ActionListener<KibanaEnrollmentResponse> listener) {
+
+        final KeyConfig keyConfig = sslService.getHttpTransportSSLConfiguration().keyConfig();
+        if (keyConfig instanceof StoreKeyConfig == false) {
+            listener.onFailure(new ElasticsearchException(
+                "Unable to enroll kibana instance. Elasticsearch node HTTP layer SSL configuration is not configured with a keystore"));
+            return;
+        }
+        List<X509Certificate> caCertificates;
+        try {
+            caCertificates = ((StoreKeyConfig) keyConfig).x509Certificates(environment)
+                .stream()
+                .filter(x509Certificate -> x509Certificate.getBasicConstraints() != -1)
+                .collect(Collectors.toList());
+        } catch (Exception e) {
+            listener.onFailure(new ElasticsearchException("Unable to enroll kibana instance. Cannot retrieve CA certificate " +
+                "for the HTTP layer of the Elasticsearch node.", e));
+            return;
+        }
+        if (caCertificates.size() != 1) {
+            listener.onFailure(new ElasticsearchException(
+                "Unable to enroll kibana instance. Elasticsearch node HTTP layer SSL configuration Keystore " +
+                "[xpack.security.http.ssl.keystore] doesn't contain a single PrivateKey entry where the associated " +
+                "certificate is a CA certificate"));
+        } else {
+            String httpCa;
+            try {
+                httpCa = Base64.getUrlEncoder().encodeToString(caCertificates.get(0).getEncoded());
+            } catch (CertificateEncodingException cee) {
+                listener.onFailure(new ElasticsearchException(
+                    "Unable to enroll kibana instance. Elasticsearch node HTTP layer SSL configuration uses a malformed CA certificate",
+                    cee));
+                return;
+            }
+            final NodesInfoRequest nodesInfoRequest = new NodesInfoRequest().addMetric(NodesInfoRequest.Metric.HTTP.metricName());
+            client.execute(NodesInfoAction.INSTANCE, nodesInfoRequest, ActionListener.wrap(nodesInfoResponse -> {
+                final List<String> nodeList = new ArrayList<>();
+                for (NodeInfo nodeInfo : nodesInfoResponse.getNodes()) {
+                    nodeList.add(nodeInfo.getInfo(HttpInfo.class).getAddress().publishAddress().toString());
```

Comment:
Publish address only exposes a single address (per node). Hopefully the one reachable from Kibana (but no way to tell for sure).

It is safer to show all the addresses, but that is not possible if we must also support binding to `0.0.0.0` (ie instead of showing publish addresses we can show bind addresses, but bind can be `0.0.0.0` which is not routable).

So there is no perfect solution. I would choose to show bind addresses because the setup script won't use `0.0.0.0`. Your call.

Reply:
Good catch. I didn't think this through as much as I should have

> I would choose to show bind addresses because the setup script won't use 0.0.0.0

The only problem with this is docker where we bind on 0.0.0.0 actually. I 'll think this through on Monday and get back with a suggestion

Reply:
Thought this through a little more:

- Kibana already can communicate with an ES node ( it just sent us a request ). The idea here is that we return the http addresses for other nodes so that it might connect to other nodes if it wants/needs. 
- bind_address is always 0.0.0.0 in docker

I think publish_address is an appropriate compromise for this. We can, however, get bind_address and return those unless it is 0.0.0.0, when we would fallback to publish_address. 

@azasypkin , @legrego ping for visibility and feedback.

Reply:
It sounds like publish address is what we would want here. While there are no guarantees that Kibana will be able to reach ES at that address, [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#modules-network-binding-publishing) claim that the publish address must be reachable by all clients (if I understand correctly).

I'm not opposed to also including the bind address in the response if it's not `0.0.0.0`, however.

Reply:
Same here, returning of `publish_address` or `publish_address` + `bind_address[]` (excluding `0.0.0.0`) sounds like a reasonable approach to me.